squeal-postgresql/src/Squeal/PostgreSQL/Session/Decode.hs

```diff
@@ -200,6 +200,16 @@ instance FromPG Strict.ByteString where
   fromPG = devalue bytea_strict
 instance FromPG Lazy.ByteString where
   fromPG = devalue bytea_lazy
+instance KnownNat n => FromPG (VarChar n) where
+  fromPG = devalue $ text_strict >>= \t ->
+    case varChar t of
+      Nothing -> throwError $ Strict.Text.pack ("source for varchar too long: " <> show (Strict.Text.length t))
+      Just x -> pure x
+instance KnownNat n => FromPG (FixChar n) where
+  fromPG = devalue $ text_strict >>= \t ->
+    case fixChar t of
+      Nothing -> throwError $ Strict.Text.pack ("source for fixchar too long: " <> show (Strict.Text.length t))
```

Comment:
"too long or too short"

Reply:
if you wanna get real fancy, you can find out the expected length from `KnownNat n` and the actual length from `Strict.Text.length t` as you did and compare them.

Reply:
Also, I recently added a `MonadFail` instance with `fail = throwError . fromString` so you can use `fail`